File: daro_cad/geom.py
"""Geometrische Grundfunktionen fuer DARO-CAD.

Alle Koordinaten sind Modellkoordinaten in Millimetern, Y-Achse zeigt nach oben
(CAD-Konvention).  Winkel werden in Grad gefuehrt und gegen den Uhrzeigersinn
gemessen -- wie in DXF.
"""

from __future__ import annotations

import math
from typing import Iterable, List, Sequence, Tuple

Point = Tuple[float, float]

EPS = 1e-9


def add(a: Point, b: Point) -> Point:
    return (a[0] + b[0], a[1] + b[1])


def sub(a: Point, b: Point) -> Point:
    return (a[0] - b[0], a[1] - b[1])


def mul(a: Point, s: float) -> Point:
    return (a[0] * s, a[1] * s)


def dot(a: Point, b: Point) -> float:
    return a[0] * b[0] + a[1] * b[1]


def cross(a: Point, b: Point) -> float:
    return a[0] * b[1] - a[1] * b[0]


def length(a: Point) -> float:
    return math.hypot(a[0], a[1])


def dist(a: Point, b: Point) -> float:
    return math.hypot(b[0] - a[0], b[1] - a[1])


def normalize(a: Point) -> Point:
    n = length(a)
    if n < EPS:
        return (0.0, 0.0)
    return (a[0] / n, a[1] / n)


def perp(a: Point) -> Point:
    """Um 90 Grad gegen den Uhrzeigersinn gedreht."""
    return (-a[1], a[0])


def lerp(a: Point, b: Point, t: float) -> Point:
    return (a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t)


def rotate(p: Point, angle_deg: float, origin: Point = (0.0, 0.0)) -> Point:
    a = math.radians(angle_deg)
    ca, sa = math.cos(a), math.sin(a)
    dx, dy = p[0] - origin[0], p[1] - origin[1]
    return (origin[0] + dx * ca - dy * sa, origin[1] + dx * sa + dy * ca)


def angle_of(a: Point) -> float:
    """Winkel des Vektors in Grad, normiert auf [0, 360)."""
    return norm_angle(math.degrees(math.atan2(a[1], a[0])))


def norm_angle(deg: float) -> float:
    deg = math.fmod(deg, 360.0)
    if deg < 0:
        deg += 360.0
    return deg


def polar(origin: Point, angle_deg: float, radius: float) -> Point:
    a = math.radians(angle_deg)
    return (origin[0] + math.cos(a) * radius, origin[1] + math.sin(a) * radius)


def arc_sweep(start_deg: float, end_deg: float) -> float:
    """Ueberstrichener Winkel eines Bogens (immer gegen den Uhrzeigersinn)."""
    sweep = norm_angle(end_deg - start_deg)
    if sweep < EPS:
        sweep = 360.0
    return sweep


def arc_point(center: Point, radius: float, angle_deg: float) -> Point:
    return polar(center, angle_deg, radius)


def arc_contains_angle(start_deg: float, end_deg: float, angle_deg: float) -> bool:
    """Liegt der Winkel auf dem Bogen von start nach end (CCW)?"""
    sweep = arc_sweep(start_deg, end_deg)
    rel = norm_angle(angle_deg - start_deg)
    return rel <= sweep + 1e-7


def flatten_arc(center: Point, radius: float, start_deg: float, end_deg: float,
                max_sagitta: float = 0.05, min_segments: int = 4) -> List[Point]:
    """Bogen in einen Polygonzug zerlegen.

    ``max_sagitta`` begrenzt den Stichmass-Fehler in mm, sodass Exporte auch bei
    grossen Radien glatt aussehen.
    """
    sweep = arc_sweep(start_deg, end_deg)
    if radius <= EPS:
        return [center, center]
    ratio = max(0.0, min(1.0, 1.0 - max_sagitta / radius))
    step = math.degrees(2.0 * math.acos(ratio)) if ratio < 1.0 else 5.0
    step = max(step, 0.5)
    n = max(min_segments, int(math.ceil(sweep / step)))
    return [arc_point(center, radius, start_deg + sweep * i / n) for i in range(n + 1)]


def bulge_to_arc(a: Point, b: Point, bulge: float):
    """DXF-Bulge in Kreisbogen umrechnen.

    Rueckgabe ``(center, radius, start_deg, end_deg)``; bei negativem Bulge
    laeuft der Bogen im Uhrzeigersinn, deshalb werden Start und Ende getauscht,
    damit die CCW-Konvention erhalten bleibt.
    """
    chord = dist(a, b)
    if chord < EPS or abs(bulge) < EPS:
        return None
    theta = 4.0 * math.atan(bulge)          # ueberstrichener Winkel, vorzeichenbehaftet
    radius = chord / (2.0 * math.sin(abs(theta) / 2.0))
    mid = lerp(a, b, 0.5)
    h = radius * math.cos(theta / 2.0)
    direction = normalize(perp(sub(b, a)))
    center = add(mid, mul(direction, h if bulge > 0 else -h))
    sa = angle_of(sub(a, center))
    ea = angle_of(sub(b, center))
    if bulge < 0:
        sa, ea = ea, sa
    return center, radius, sa, ea


# --------------------------------------------------------------------------
# Schnittpunkte -- Basis fuer Fangpunkte, Trimmen und Abrunden
# --------------------------------------------------------------------------

def line_line_intersection(a1: Point, a2: Point, b1: Point, b2: Point,
                           segment: bool = True):
    """Schnittpunkt zweier Strecken bzw. Geraden, ``None`` bei Parallelitaet."""
    r = sub(a2, a1)
    s = sub(b2, b1)
    denom = cross(r, s)
    if abs(denom) < EPS:
        return None
    t = cross(sub(b1, a1), s) / denom
    u = cross(sub(b1, a1), r) / denom
    if segment and not (-1e-9 <= t <= 1 + 1e-9 and -1e-9 <= u <= 1 + 1e-9):
        return None
    return add(a1, mul(r, t))


def line_circle_intersection(a: Point, b: Point, center: Point, radius: float,
                             segment: bool = True) -> List[Point]:
    d = sub(b, a)
    f = sub(a, center)
    aa = dot(d, d)
    if aa < EPS:
        return []
    bb = 2 * dot(f, d)
    cc = dot(f, f) - radius * radius
    disc = bb * bb - 4 * aa * cc
    if disc < 0:
        return []
    disc = math.sqrt(disc)
    out = []
    for t in ((-bb - disc) / (2 * aa), (-bb + disc) / (2 * aa)):
        if segment and not (-1e-9 <= t <= 1 + 1e-9):
            continue
        out.append(add(a, mul(d, t)))
    return out


def circle_circle_intersection(c1: Point, r1: float, c2: Point, r2: float) -> List[Point]:
    d = dist(c1, c2)
    if d < EPS or d > r1 + r2 + EPS or d < abs(r1 - r2) - EPS:
        return []
    a = (r1 * r1 - r2 * r2 + d * d) / (2 * d)
    h2 = r1 * r1 - a * a
    h = math.sqrt(max(0.0, h2))
    base = add(c1, mul(sub(c2, c1), a / d))
    if h < EPS:
        return [base]
    off = mul(perp(normalize(sub(c2, c1))), h)
    return [add(base, off), sub(base, off)]


def closest_point_on_segment(p: Point, a: Point, b: Point) -> Point:
    d = sub(b, a)
    aa = dot(d, d)
    if aa < EPS:
        return a
    t = max(0.0, min(1.0, dot(sub(p, a), d) / aa))
    return add(a, mul(d, t))


# --------------------------------------------------------------------------
# Polygone
# --------------------------------------------------------------------------

def signed_area(pts: Sequence[Point]) -> float:
    s = 0.0
    n = len(pts)
    for i in range(n):
        x1, y1 = pts[i]
        x2, y2 = pts[(i + 1) % n]
        s += x1 * y2 - x2 * y1
    return s / 2.0


def is_ccw(pts: Sequence[Point]) -> bool:
    return signed_area(pts) > 0


def point_in_polygon(p: Point, pts: Sequence[Point]) -> bool:
    """Ray-Casting; Punkte auf dem Rand gelten als innen."""
    x, y = p
    inside = False
    n = len(pts)
    for i in range(n):
        x1, y1 = pts[i]
        x2, y2 = pts[(i + 1) % n]
        if abs(cross(sub((x2, y2), (x1, y1)), sub(p, (x1, y1)))) < 1e-9:
            if min(x1, x2) - 1e-9 <= x <= max(x1, x2) + 1e-9 and \
               min(y1, y2) - 1e-9 <= y <= max(y1, y2) + 1e-9:
                return True
        if (y1 > y) != (y2 > y):
            xi = x1 + (y - y1) / (y2 - y1) * (x2 - x1)
            if xi > x:
                inside = not inside
    return inside


def bbox(points: Iterable[Point]):
    pts = list(points)
    if not pts:
        return None
    xs = [p[0] for p in pts]
    ys = [p[1] for p in pts]
    return (min(xs), min(ys), max(xs), max(ys))


def dedupe(pts: Sequence[Point], tol: float = 1e-7) -> List[Point]:
    out: List[Point] = []
    for p in pts:
        if not out or dist(out[-1], p) > tol:
            out.append(p)
    return out

File: web/js/geom.js
// Geometrie-Grundlagen der Zeichenflaeche.
// Spiegelt daro_cad/geom.py -- gleiche Konventionen: Millimeter, Y nach oben,
// Winkel in Grad gegen den Uhrzeigersinn.

export const EPS = 1e-9;

export const add = (a, b) => [a[0] + b[0], a[1] + b[1]];
export const sub = (a, b) => [a[0] - b[0], a[1] - b[1]];
export const mul = (a, s) => [a[0] * s, a[1] * s];
export const dot = (a, b) => a[0] * b[0] + a[1] * b[1];
export const cross = (a, b) => a[0] * b[1] - a[1] * b[0];
export const len = (a) => Math.hypot(a[0], a[1]);
export const dist = (a, b) => Math.hypot(b[0] - a[0], b[1] - a[1]);
export const perp = (a) => [-a[1], a[0]];
export const lerp = (a, b, t) => [a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t];
export const eq = (a, b, tol = 1e-7) => dist(a, b) <= tol;

export function normalize(a) {
  const n = len(a);
  return n < EPS ? [0, 0] : [a[0] / n, a[1] / n];
}

export function rotate(p, deg, origin = [0, 0]) {
  const a = deg * Math.PI / 180, c = Math.cos(a), s = Math.sin(a);
  const dx = p[0] - origin[0], dy = p[1] - origin[1];
  return [origin[0] + dx * c - dy * s, origin[1] + dx * s + dy * c];
}

export function normAngle(deg) {
  let d = deg % 360;
  if (d < 0) d += 360;
  return d;
}

export const angleOf = (a) => normAngle(Math.atan2(a[1], a[0]) * 180 / Math.PI);

export function polar(origin, deg, radius) {
  const a = deg * Math.PI / 180;
  return [origin[0] + Math.cos(a) * radius, origin[1] + Math.sin(a) * radius];
}

export function arcSweep(start, end) {
  const s = normAngle(end - start);
  return s < EPS ? 360 : s;
}

export const arcPoint = (c, r, deg) => polar(c, deg, r);

export function arcContains(start, end, deg) {
  return normAngle(deg - start) <= arcSweep(start, end) + 1e-7;
}

export function flattenArc(c, r, start, end, steps = 0) {
  const sweep = arcSweep(start, end);
  const n = steps || Math.max(6, Math.ceil(sweep / 6));
  const out = [];
  for (let i = 0; i <= n; i++) out.push(arcPoint(c, r, start + sweep * i / n));
  return out;
}

// DXF-Bulge -> Kreisbogen {c, r, start, end}
export function bulgeToArc(a, b, bulge) {
  const chord = dist(a, b);
  if (chord < EPS || Math.abs(bulge) < EPS) return null;
  const theta = 4 * Math.atan(bulge);
  const r = chord / (2 * Math.sin(Math.abs(theta) / 2));
  const mid = lerp(a, b, 0.5);
  const h = r * Math.cos(theta / 2);
  const dir = normalize(perp(sub(b, a)));
  const c = add(mid, mul(dir, bulge > 0 ? h : -h));
  let sa = angleOf(sub(a, c)), ea = angleOf(sub(b, c));
  if (bulge < 0) { const t = sa; sa = ea; ea = t; }
  return { c, r, start: sa, end: ea };
}

export function arcToBulge(c, r, start, end) {
  const sweep = arcSweep(start, end);
  return Math.tan(sweep * Math.PI / 180 / 4);
}

// -- Schnittpunkte ---------------------------------------------------------

export function lineLine(a1, a2, b1, b2, segment = true) {
  const r = sub(a2, a1), s = sub(b2, b1);
  const denom = cross(r, s);
  if (Math.abs(denom) < EPS) return null;
  const t = cross(sub(b1, a1), s) / denom;
  const u = cross(sub(b1, a1), r) / denom;
  if (segment && !(t >= -1e-9 && t <= 1 + 1e-9 && u >= -1e-9 && u <= 1 + 1e-9)) return null;
  return add(a1, mul(r, t));
}

export function lineCircle(a, b, c, r, segment = true) {
  const d = sub(b, a), f = sub(a, c);
  const aa = dot(d, d);
  if (aa < EPS) return [];
  const bb = 2 * dot(f, d);
  const cc = dot(f, f) - r * r;
  let disc = bb * bb - 4 * aa * cc;
  if (disc < 0) return [];
  disc = Math.sqrt(disc);
  const out = [];
  for (const t of [(-bb - disc) / (2 * aa), (-bb + disc) / (2 * aa)]) {
    if (segment && (t < -1e-9 || t > 1 + 1e-9)) continue;
    out.push(add(a, mul(d, t)));
  }
  return out;
}

export function circleCircle(c1, r1, c2, r2) {
  const d = dist(c1, c2);
  if (d < EPS || d > r1 + r2 + EPS || d < Math.abs(r1 - r2) - EPS) return [];
  const a = (r1 * r1 - r2 * r2 + d * d) / (2 * d);
  const h = Math.sqrt(Math.max(0, r1 * r1 - a * a));
  const base = add(c1, mul(sub(c2, c1), a / d));
  if (h < EPS) return [base];
  const off = mul(perp(normalize(sub(c2, c1))), h);
  return [add(base, off), sub(base, off)];
}

export function closestOnSegment(p, a, b) {
  const d = sub(b, a), aa = dot(d, d);
  if (aa < EPS) return a;
  const t = Math.max(0, Math.min(1, dot(sub(p, a), d) / aa));
  return add(a, mul(d, t));
}

export function distToSegment(p, a, b) {
  return dist(p, closestOnSegment(p, a, b));
}

export function distToArc(p, c, r, start, end) {
  const ang = angleOf(sub(p, c));
  if (arcContains(start, end, ang)) return Math.abs(dist(p, c) - r);
  return Math.min(dist(p, arcPoint(c, r, start)), dist(p, arcPoint(c, r, end)));
}

// -- Polygone --------------------------------------------------------------

export function signedArea(pts) {
  let s = 0;
  for (let i = 0; i < pts.length; i++) {
    const a = pts[i], b = pts[(i + 1) % pts.length];
    s += a[0] * b[1] - b[0] * a[1];
  }
  return s / 2;
}

export function pointInPolygon(p, pts) {
  let inside = false;
  for (let i = 0, j = pts.length - 1; i < pts.length; j = i++) {
    const a = pts[i], b = pts[j];
    if ((a[1] > p[1]) !== (b[1] > p[1])) {
      const x = a[0] + (p[1] - a[1]) / (b[1] - a[1]) * (b[0] - a[0]);
      if (x > p[0]) inside = !inside;
    }
  }
  return inside;
}

export function bbox(points) {
  if (!points.length) return null;
  let [x0, y0] = points[0], x1 = x0, y1 = y0;
  for (const p of points) {
    if (p[0] < x0) x0 = p[0];
    if (p[1] < y0) y0 = p[1];
    if (p[0] > x1) x1 = p[0];
    if (p[1] > y1) y1 = p[1];
  }
  return [x0, y0, x1, y1];
}

export const bboxOverlap = (a, b) =>
  a && b && a[0] <= b[2] && a[2] >= b[0] && a[1] <= b[3] && a[3] >= b[1];

export const bboxInside = (inner, outer) =>
  inner && outer && inner[0] >= outer[0] && inner[2] <= outer[2] &&
  inner[1] >= outer[1] && inner[3] <= outer[3];

export function round(value, step) {
  return step > 0 ? Math.round(value / step) * step : value;
}
